### PR TITLE
feat: calculation of deadlines by segment

### PR DESCRIPTION
**Pull request by kobayu858**
## Description

- Calculation of deadlines by segment.

## Related links

- original paper [OPEN_LINK](https://drive.google.com/file/d/1YH-MAWsw1VAvxqco4Beg3njQ9ykG-sb-/view?usp=drive_link) p5, p6
- paper summary [EmbIV_INTERNAL_LINK](https://docs.google.com/presentation/d/1rabhhA9ziIcBjvT8XPLue78XEWcKMvH4/edit?usp=drive_link&ouid=114908281052440997053&rtpof=true&sd=true) p27 - p34
- design specifications [EmbIV INTERNAL LINK](https://docs.google.com/presentation/d/14YO4wHY1QqGOjoY9_g4djvtTaBXV0RCj/edit?usp=drive_link&ouid=114908281052440997053&rtpof=true&sd=true) p4, p5

## Pre-review checklist for the PR author

- [x] I've confirmed the [self-checklist](https://docs.google.com/spreadsheets/d/1p5KMcm752iCySHJz0xL1qf6fcLDKprbo/edit#gid=863618628).
- [x] The PR is ready for review.

## In-review checklist for the PR reviewers

The PR reviewers **must** check the checkboxes below before approval.

- [ ] The PR has been properly tested.
- [ ] The PR has been reviewed.

## Post-review checklist for the PR author

The PR author **must** check the checkboxes below before merging.

- [ ] There are no open discussions.
- [ ] The PR is ready for merge.

After all checkboxes are checked, anyone who has write access can merge the PR.

### CODE REVIEW

File: 2014_TPDS_basic_decomposition_based_algorithm/src/handle_segment.rs [hunk 1]
```diff
@@ -16,9 +16,11 @@ pub struct Segment {
     pub nodes: Vec<NodeData>,
     pub begin_range: i32,
     pub end_range: i32,
-    pub deadline: f32, //TODO: use
+    pub deadline: f32,
     pub classification: Option<SegmentClassification>,
-    pub parallel_degree: i32,
+    pub execution_requirement: i32, // end_range - begin_range
+    pub parallel_degree: i32,       // number of nodes in the segment
+    pub volume: i32,                // execution_requirement * volume
```
Comment (atsushi421) [inline 2014_TPDS_basic_decomposition_based_algorithm/src/handle_segment.rs:23]:
```suggestion
    pub volume: i32,                // execution_requirement * nodes.len()
```

File: 2014_TPDS_basic_decomposition_based_algorithm/src/handle_segment.rs [hunk 2]
```diff
@@ -104,6 +111,58 @@ fn classify_dag(
...
+            }
+        }
+        DAGClassification::Light => {
+            for segment in segments {
+                segment.deadline = (period / crit_path_len) * segment.execution_requirement as f32;
+            }
+        }
+        DAGClassification::Mixture => {
+            let (mut heavy_segment_volume, mut light_segment_length) = (0, 0);
+            for segment in segments.iter_mut() {
```
Comment (atsushi421) [inline 2014_TPDS_basic_decomposition_based_algorithm/src/handle_segment.rs:135]:
```suggestion
            for segment in segments.iter() {
```

**Review by atsushi421 — CHANGES_REQUESTED** [review]

File: 2014_TPDS_basic_decomposition_based_algorithm/src/handle_segment.rs (hunk 1 above)
Comment (kobayu858) [reply 2014_TPDS_basic_decomposition_based_algorithm/src/handle_segment.rs:23]: Correction omission. Thanks for pointing this out.

File: 2014_TPDS_basic_decomposition_based_algorithm/src/handle_segment.rs (hunk 2 above)
Comment (kobayu858) [reply 2014_TPDS_basic_decomposition_based_algorithm/src/handle_segment.rs:135]: Changed to immutable.

**Review by atsushi421 — APPROVED** [review]
LGTM

**Pull request merged**